jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientProxyProtocolTest.java

```diff
@@ -0,0 +1,240 @@
+import org.eclipse.jetty.client.api.Destination;
+import org.eclipse.jetty.http.HttpHeader;
+import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.http.MimeTypes;
+import org.eclipse.jetty.io.EndPoint;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.HttpConnectionFactory;
+import org.eclipse.jetty.server.ProxyConnectionFactory;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.util.thread.QueuedThreadPool;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+
+import static org.eclipse.jetty.client.ProxyProtocolClientConnectionFactory.V1;
+import static org.eclipse.jetty.client.ProxyProtocolClientConnectionFactory.V2;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class HttpClientProxyProtocolTest
+{
+    private Server server;
+    private ServerConnector connector;
+    private HttpClient client;
+
+    private void startServer(Handler handler) throws Exception
+    {
+        QueuedThreadPool serverThreads = new QueuedThreadPool();
+        serverThreads.setName("server");
+        server = new Server(serverThreads);
+        HttpConnectionFactory http = new HttpConnectionFactory();
+        ProxyConnectionFactory proxy = new ProxyConnectionFactory(http.getProtocol());
+        connector = new ServerConnector(server, 1, 1, proxy, http);
+        server.addConnector(connector);
+        server.setHandler(handler);
+        server.start();
+    }
+
+    private void startClient() throws Exception
+    {
+        QueuedThreadPool clientThreads = new QueuedThreadPool();
+        clientThreads.setName("client");
+        client = new HttpClient();
+        client.setExecutor(clientThreads);
+        client.setRemoveIdleDestinations(false);
+        client.start();
+    }
+
+    @AfterEach
+    public void dispose() throws Exception
+    {
+        if (server != null)
+            server.stop();
+        if (client != null)
+            client.stop();
+    }
+
+    @Test
+    public void testClientProxyProtocolV1() throws Exception
+    {
+        startServer(new EmptyServerHandler()
+        {
+            @Override
+            protected void service(String target, Request jettyRequest, HttpServletRequest request, HttpServletResponse response) throws IOException
+            {
+                response.setContentType(MimeTypes.Type.TEXT_PLAIN.asString());
+                response.getOutputStream().print(request.getRemotePort());
+            }
+        });
+        startClient();
+
+        int serverPort = connector.getLocalPort();
+
+        int clientPort = ThreadLocalRandom.current().nextInt(1024, 65536);
+        V1.Info info = new V1.Info("TCP4", "127.0.0.1", clientPort, "127.0.0.1", serverPort);
```

Comment:
This example makes no sense to me?   The user should not be dealing with IP addresses and ports like this.  It is not entirely clean which client this is talking about -  specifically it looks like I need to know my own client port before it has even been allocated (which it is not as I guess it is the port that we are forwarding??).
Why must the user specify information that is in the request anyway, such as the server and serverPort?  Does it have to be server IP or can it be server name?  What about servers with multiple IPs and DNS balancing?